### PR TITLE
plumb c10::nullopt through python binding layer when dtype isn't specified

### PR DESCRIPTION
**Pull request by bdhirsh**
Fixes https://github.com/pytorch/pytorch/issues/63356.

Specifically, this code example works now (successfully plumbs `c10::nullopt` for the dtype argument to `torch.empty_quantized`, allowing it to default to the dtype of the passed-in tensor instead).
```
import torch
x = torch.randn(3)
 xq = torch.quantize_per_tensor(x, 0, 1, torch.quint8)
xq2 = torch.empty_quantized([4], xq)
```

The previous generated code for `THPVariable_empty_quantized` that created the TensorOptions argument looked like this:
```
const auto options = TensorOptions()
    .dtype(_r.scalartype(3)) // this immediately substitutes float if the argument is None
    .device(_r.device(5)) // this immediately substitutes cpu if the argument is None
    .layout(_r.layoutOptional(4))
    .requires_grad(_r.toBool(7))
    .pinned_memory(_r.toBool(6));
```
and it now looks like this:
```
const auto options = TensorOptions()
    .dtype(_r.scalartypeOptional(3))
    .device(_r.device(5)) // Not fixing this, too BC-breaking. See comments in the PR
    .layout(_r.layoutOptional(4))
    .requires_grad(_r.toBool(7))
    .pinned_memory(_r.toBool(6));
```

In the python codegen, we manually construct the list of `PythonArgument`'s for operators that contain a `TensorOptions` argument. Long term, we probably want to unify the two (and use a generic function to map from `Argument` -> `PythonArgument`, but I opted for a more tactical fix that causes fewer BC-breaking changes.

This is still technically BC-breaking, but I think it should only impact ops that:
- take in both TensorOptions and a Tensor argument
- don't have "new" or "like" in their name (i.e. `ones_like`, etc)

And it looks like the only ops that fit that criteria are the two relevant quantized factory ops (which we want to fix the behavior of anyway):
- _empty_per_channel_affine_quantized
- empty_quantized

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63906 plumb c10::nullopt Device through python binding layer, match TensorOptions defaulting logic with python binding defaulting logic
* **#63364 plumb c10::nullopt through python binding layer when dtype isn't specified**

Differential Revision: [D30354319](https://our.internmc.facebook.com/intern/diff/D30354319)

### CODE REVIEW

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30354319).

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30354319).

**Comment by bdhirsh** [comment]
the test failures are suspicious but I couldn't repro them locally. Rebasing to see if they still show up

**Comment by bdhirsh** [comment]
It looks like I need to patch the `{hann|hamming|bartlett|blackman}_window` kernels to keep the same behavior. Previously, they'd all call into `at::arange()` with whatever dtype was passed in. That used to be `float` (which was defaulted at the python binding layer), but now it's `c10::nullopt` -  which now causes the default output dtype to be `long`, and breaks stuff because they all try to call `.mul_(<float_tensor>)`.

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30354319).

**Comment by bdhirsh** [comment]
oof. I wanted to fix both `dtype` and `device`, but `device` seems more BC-breaking: the defaulted device used in the python binding layer is pulled from `torch::tensors::set_default_tensor_type`. But the default device that `TensorOptions` is just `kCPU`, so this breaks a bunch of tests.

There's a scale on what type of fix I can make, ranging from "super tactical, minimize BC-breakage" to "unify the python codegen, very BC-breaking". I'm hoping to fix the specific issue with `empty_quantized` mentioned in the issue, so I'm leaning towards being more tactical. The most tactical thing that I could do is special-case `empty_quantized` in the codegen, which I really don't want to do. So for now I'll try "only" fixing `dtype` in the codegen but not device, and see if lots of tests still break.

**Comment by HDCharles** [comment]
if you take the dtype only approach, could you tactically fix device for empty_quantized as well? 

its significantly less important (the main use case of the function is when device is specified) but it'd be unexpected for gpu tensors to suddenly appear on cpu when calling that function from python.

**Comment by ezyang** [comment]
I remember that I tried this a while ago in https://github.com/pytorch/pytorch/pull/33826 and I think I ran into the same problem you did, which is that the Python-side and C++-side defaults aren't consistent. In fact, you technically have the same problem with dtype, because there is also set_default_dtype (and you probably just haven't gotten to the tests that tickle it).  I'd say, either just special case this function, or start mirroring these Python-side defaults into C++...

**Comment by bdhirsh** [comment]
Actually, I'm pretty sure that both places (python_arg_parser.h and TensorOptions.h) pull the default dtype from the same place:  https://github.com/pytorch/pytorch/blob/30e1c74dc19ae2b622b46ebcdb7972c42775ac80/c10/core/DefaultDtype.cpp#L5.  Gonna try it and see if any tests break.

I'll also include the special casing  of `empty_quantized`'s `device` argument. And if I do hit issues trying to plumb just the `dtype` argument, I'll special case that function for `dtype` too.

**Comment by ezyang** [comment]
Need more review on this?

**Comment by bdhirsh** [comment]
> Need more review on this?

Yep, I think this is ready (but the subsequent PR's aren't). This PR fixes the plumbing just for dtype, and I'm hoping to do later ones for device (and layout?) afterwards.

It looks like the only failure right now is xla, and I'm working on a pt/xla-side fix for it (pt/xla's type promotion for `div.Scalar` is broken)

File: aten/src/ATen/native/TensorFactories.cpp [hunk 1]
```diff
@@ -1320,7 +1320,7 @@ Tensor kaiser_window(
   if (periodic) {
     window_length += 1;
   }
-  auto initial = at::arange(window_length, options);
+  auto initial = at::arange(window_length, typeMetaToScalarType(options.dtype()), options.layout(), options.device(), options.pinned_memory());
```
Comment (ezyang) [inline aten/src/ATen/native/TensorFactories.cpp:1323]: at::arange TensorOptions should have kept working no?

File: torch/csrc/utils/out_types.cpp [hunk 2]
```diff
@@ -6,29 +6,28 @@ namespace utils {
...
-        "scalar type ", scalarType_arg,
+        "scalar type ", *scalarType,
         " does not match scalar type of out parameter (", result.scalar_type(), ")");
   }
   if (layout && result.layout() != *layout) {
     AT_ERROR(
         "layout ", *layout,
         " does not match layout of out parameter (", result.layout(), ")");
   }
+  auto device_type_arg = device_is_none ? result.device().type() : device.type();
```
Comment (ezyang) [inline torch/csrc/utils/out_types.cpp:30]: unchanged

**Review by ezyang — APPROVED** [review]
If it works!

**Pull request closed without merging**